Content/Items/Misc/Weapons.Soilgun.cs

```diff
@@ -0,0 +1,1488 @@
+            return false;
+        }
+
+        public override void AddRecipes()
+        {
+            CreateRecipe().AddIngredient(ItemID.DirtRod).AddIngredient(ItemID.Sandgun).AddIngredient(ItemID.DirtBlock, 45).AddTile(TileID.WorkBenches).Register();
+        }
+
+        public override void UpdateInventory(Player player)
+        {
+            //TODO: make this prioritize top left -> bottom right, instead of bottom right -> top left, like actual ammo
+            for (int i = 0; i < player.inventory.Length; i++)
+            {
+                Item item = player.inventory[i];
+
+                if (ValidSoils.Contains(item.type))
+                {
+                    SoilgunPlayer modplayer = player.GetModPlayer<SoilgunPlayer>();
+
+                    modplayer.SoilItem = item;
+                    modplayer.HasSoilItem = true;
+                }
+            }
+        }
+
+        public override void ModifyTooltips(List<TooltipLine> tooltips)
+        {
+            if (Main.LocalPlayer.GetModPlayer<SoilgunPlayer>().SoilItem == null)
+                return;
+
+            TooltipLine AmmoLine = new TooltipLine(StarlightRiver.Instance, "AmmoLineToolTip", $"Current Ammo: [i:{Main.LocalPlayer.GetModPlayer<SoilgunPlayer>().SoilItem.type}]{Main.LocalPlayer.GetModPlayer<SoilgunPlayer>().SoilItem.stack}");
+            tooltips.Add(AmmoLine);
+        }
+        //TODO make this look not bad
+        public override void PostDrawInInventory(SpriteBatch spriteBatch, Vector2 position, Rectangle frame, Color drawColor, Color itemColor, Vector2 origin, float scale)
+        {
+            if (Main.playerInventory || Main.LocalPlayer.GetModPlayer<SoilgunPlayer>().SoilItem == null)
+                return;
+
+            Utils.DrawBorderStringFourWay(spriteBatch, FontAssets.ItemStack.Value, $"{ Main.LocalPlayer.GetModPlayer<SoilgunPlayer>().SoilItem.stack}", position.X, position.Y + 15, Color.White, Color.Black, origin, (scale + 0.2f) * Main.UIScale);
+        }
+
+        //increases damage depending on type of soil, like ammos
+        public int GetSoilDamage(Player player)
+        {
+            if (player.GetModPlayer<SoilgunPlayer>().SoilItem.type == Mod.Find<ModItem>("VitricSandItem").Type)
+                return 8;
+            switch (player.GetModPlayer<SoilgunPlayer>().SoilItem.type)
+            {
+                case ItemID.SandBlock:
+                    return 2;
+                case ItemID.CrimsandBlock:
+                    return 4;
+                case ItemID.EbonsandBlock:
+                    return 4;
+                case ItemID.PearlsandBlock:
+                    return 15;
+                case ItemID.DirtBlock:
+                    return 2;
+                case ItemID.SiltBlock:
+                    return 3;
+                case ItemID.SlushBlock:
+                    return 3;
+            }
+            return 0;
+        }
+    }
+
+    class SoilgunPlayer : ModPlayer
```

Comment:
It may be worthwhile looking into solutions using TModLoader ammo hooks or even custom hooks to pick ammo over this.